ludwig/models/abstractmodel.py

```diff
@@ -0,0 +1,242 @@
+        output_features_def = topological_sort_feature_dependencies(output_features_def)
+        output_features = {}
+
+        for output_feature_def in output_features_def:
+            # TODO(Justin): Check that the semantics of input_size align with what the combiner's output shape returns
+            # for seq2seq.
+            output_feature_def["input_size"] = combiner.output_shape[-1]
+            output_feature = cls.build_single_output(output_feature_def, output_features)
+            output_features[output_feature_def[NAME]] = output_feature
+
+        return output_features
+
+    @staticmethod
+    def build_single_output(
+        output_feature_def: Dict[str, Any], output_features: Dict[str, OutputFeature]
+    ) -> OutputFeature:
+        """Builds a single output feature from the output feature definition."""
+        logger.debug(f"Output {output_feature_def[TYPE]} feature {output_feature_def[NAME]}")
+
+        output_feature_class = get_from_registry(output_feature_def[TYPE], output_type_registry)
+        output_feature_obj = output_feature_class(output_feature_def, output_features)
+
+        return output_feature_obj
+
+    def get_model_inputs(self):
+        inputs = {
+            input_feature_name: input_feature.create_sample_input()
+            for input_feature_name, input_feature in self.input_features.items()
+        }
+        return inputs
+
+    # Return total number of parameters in model
+    def get_model_size(self) -> int:
+        model_tensors = self.collect_weights()
+        total_size = 0
+        for tnsr in model_tensors:
+            total_size += tnsr[1].detach().cpu().numpy().size
+        return total_size
+
+    def to_torchscript(self):
+        self.eval()
+        model_inputs = self.get_model_inputs()
+        # We set strict=False to enable dict inputs and outputs.
+        return torch.jit.trace(self, model_inputs, strict=False)
+
+    def save_torchscript(self, save_path):
+        traced = self.to_torchscript()
+        traced.save(save_path)
+
+    @property
+    def input_shape(self):
+        # TODO(justin): Remove dummy implementation. Make input_shape and output_shape functions.
+        return torch.Size([1, 1])
+
+    @abstractmethod
+    def forward(
+        self,
+        inputs: Union[
+            Dict[str, torch.Tensor], Dict[str, np.ndarray], Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]
+        ],
+        mask=None,
+    ) -> Dict[str, torch.Tensor]:
+        """Forward pass of the model.
+
+        Args:
+            inputs: Inputs to the model. Can be a dictionary of input names to
+                input tensors or a tuple of (inputs, targets) where inputs is
+                a dictionary of input names to input tensors and targets is a
+                dictionary of target names to target tensors.
+            mask: A mask for the inputs.
+
+        Returns:
+            A dictionary of output {feature name}::{tensor_name} -> output tensor.
+        """
```

Comment:
Should we raise a `NotImplementedError` if this method isn't implemented?

Reply:
By using ABCMeta metaclass, the current behavior is as follows if `forward()` is not implemented:

```
>>> from ludwig.models.base import BaseModel
>>> class Test(BaseModel):
...   pass
... 
>>> Test()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Can't instantiate abstract class Test with abstract methods forward
```